pep-9999.rst

```diff
@@ -0,0 +1,257 @@
+PEP: 9999
+Title: Package index configuration standardization
+Author: Fredrik Averpil <fredrik at averpil.com>
+Status: Draft
+Type: Informational
+Created: 05-May-2021
+Post-History: <dates of postings to python-ideas and/or python-dev>
+
+
+Abstract
+========
+
+This PEP standardizes configuration of PyPI repositories.
+
+
+Motivation
+==========
+
+Workflow tools come up with their own configuration mechanisms for PyPI repositories, as there is no standard for this. This causes fragmentation and pain points.
+
+Examples of such configuration files:
+
+* Pip's ``pip.conf`` and ``pip.ini``
+* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
+* Poetry's ``config.toml`` and ``auth.toml``
+
+
+Rationale
+=========
+
+While this PEP aim to come up with a unified PyPI repository configuration, thought has been given so to be able to extend the configuration later. For example, PyPI repositories are assumed in this PEP, but a ``type`` attribute could be introduced which would default to ``pypi`` down the line, to allow for other package repository types. Any tool could also start taking advantage of the fact that the configuration files and their discovery is a standard, and use this for their own configuration.
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for PyPI repository configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file::
+
+
+    [repositories]
+
+    [repostiories.pypi]
+    url = "https://pypi.org/simple/"
+
+    [repostiories.myrepo1]
+    url = "http://myserver2.org/simple"
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+A repository has the following attributes::
+
+    ==================  ===========  =============================================
+    Attribute           Type         Description
+    ==================  ===========  =============================================
+    final               bool         Make it impossible to override/remove the repository configuration (default: false)
+    priority            int          Order of priority, sorted descending (default: 0)
+    proxy_url           str          The URL to the proxy server, can contain a port
```

Comment:
Would be great if proxy config is a nested table rather than attributes as they themselves.

Reply:
Sure, why not.

What attributes would such a config have, in your opinion?
Besides `proxy_url` I mean.

Reply:
everything proxy related; credentials and I think in some cases the certs also apply i need to look into that a bit. I am also curious how this interacts with system configuration. Are we implicitly saying that any config here will take precedence over system config? Most cases I have encountered (atleast with enterprise clients) have been interaction with system proxy. Having this configuration required in tool configuration often caused more pain than otherwise. But this might just be me being paranoid as well. :)

Reply:
> Are we implicitly saying that any config here will take precedence over system config?

I have to say I don't fully understand how a proxy setting would work either. It was brought up in the original discussion thread: https://discuss.python.org/t/adding-a-global-config-to-specify-package-indexes/8599/5

Reply:
>Agreed, I would love to see a standardised format that is used by all tools, for things like package indexes, network settings (proxy, certificates), credentials etc.

So. This is interesting. The comment is a bit ambiguous I'd say. The intent here might be network settings at the same level as package index config. Because, I am not sure if per repo proxy config makes any sense.  I'd say that this makes more sense, meaning cliet tools respect proxy settings if specified - this would require it to ignore system settings and also add support for things like netrc files. Sounds more like a separate PEP tbh.

(scope creep) In the same sense I'd even say that credentials should exist at the same level. And package indexes referring to them by name maybe if hostname is ambigious otherwise do a hostname match?